working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+In most situations, there is no difference between declaring the constructor above explicitly and getting the default constructor. The only distinction is that you can extract a mixin from a class if it has `Object` as superclass and it declares no non-factory constructor.
+
+### Mixin Application Forwarding Constructors
+
+A *mixin application* receives corresponding forwarding constructors for all accessible generative constructors of the superclass. The forwarding constructor has the same basename (identifer after the class name for named constructors, empty for the unnamed constructor), the same parameter signature as the superclass constructor, including default values for optional parameters, and an initializer list with a `super`-invocation to the corresponding superclass constructor which forwards each parameter as an argument to the corresponding superclass constructor parameter. The forwarding constructor is `const` if the superclass constructor is `const` *and* the mixin declares no instance variables.
+
+Since a mixin application class cannot contain static members, there is no risk of the introducing a conflict between a constructor name and a static member. 
+
+### Goal
+
+We combine the forwarding constructor feature with default constructors and automatic initialization of declared instance variables. This *unifies* default constructors and mixin application forwarding constructors *and* adds more power to the default constructor for simple classes.
+
+We expect this to happen after non-nullable types have been added to the language, so the proposal is written in terms of that, with a few guidelines for how legacy libraries work where nullability matters.
+
+## Proposal
+
+Let *C* be a class declaration with name `C` and superclass *S* with name `S` and which *declares* no constructor.
+
+It is a compile-time error if:
+
+- *C* declares a private-named and potentially non-nullable instance variable with no initializer expressions. 
+- Or in **NNBD-legacy mode**,  *C* declares a private-named and final instance variable with no initializer expression.
+
+Otherwise, for each constructor *s* of the superclass, named `S` or `S.id`, satisfying all of the following:
+
+1. If *s* is a named constructor with name `S.id` then:
+   1.  *s* is accessible ( `id` is not a private identifier or *S* is not declared in a different library), and
+   2.  *C* does not declare a static member named *id*.
+2. The *s* constructor is a generative constructor.
+3. *C* does not declare an instance variable with no initializer expression and with the same basename as a required named parameter of *s*.
+4. Either *C* does not declare any instance variables or *s* does not have any optional positional parameters (required because we can't declare a constructor with both optional position parameters and named parameters).
```

Comment:
I think this makes it a breaking change to add an optional parameter to a constructor.  Are we comfortable with that?

Reply:
Not particularly. Can we have both kinds of optionals on the same constructor? (PLEASE!!!!)
Otherwise forwarding to a non-default constructor is dead in the water. Which is sad, but maybe necessary for now.

Reply:
> Can we have both kinds of optionals on the same constructor?

YESPLZ.

Reply:
I've reeled in the forwarding a little, so we never forward optional positional parameters if adding one would be breaking. It means that you can forward optional positional superconstructor parameters, as long as you put the `super` after all other positional parameters in the subclass constructor (and have no named parameters, obviously).